Assets/Minigames/Fight/Scripts/Managers/RoomManager.cs

```diff
@@ -1,7 +1,173 @@
+
+
+            List<RoomController> roomsToInstantiate = new();
+            Vector2 direction = Vector2.zero;
+
+            // Populate our rooms to instantiate list with the determined number of random rooms
+            for (int i = 0; i < roomCount; i++)
+            {
+                roomsToInstantiate.Add(_roomSettings.GetRandomRoom());
+            }
+
+            // Generate the map
+            foreach (RoomController roomToInstantiate in roomsToInstantiate)
+            {
+                bool valid = false;
+                var room = roomToInstantiate;
+
+                Vector2 center = Vector2.zero;
+
+                // Repeat until valid placement has been found
+                while (!valid)
+                {
+                    // Select a random room to branch off of.
+                    int roomToBranchOffOf = Random.Range(0, availableRooms.Count);
+                    targetRoom = availableRooms[roomToBranchOffOf];
+
+                    // Select random direction to branch in.
+                    direction = GetRandomCardinalDirection();
+
+                    // Calculate the location the new room will spawn in.
+                    float x = (targetRoom.col.bounds.extents.x + (room.Tilemap.cellBounds.AsVector2().x / 2)) * direction.x;
+                    float y = (targetRoom.col.bounds.extents.y + (room.Tilemap.cellBounds.AsVector2().y / 2)) * direction.y;
+                    center = targetRoom.col.bounds.center.AsVector2() + new Vector2(x, y);
+
+
+                    // Abort before doing the overlap check if the chosen direction has already been used (slight performance boost).
+                    RoomConnection roomConnection = null;
+                    foreach (RoomConnection connection in targetRoom.roomConnections)
+                    {
+                        if (connection.Direction == direction)
+                        {
+                            if (connection.HasConnection)
+                            {
+                                continue;
+                            }
+                            roomConnection = connection;
+                        }
+                    }
+
+                    // Check perposed room location to ensure it won't overlap with an existing room
+                    if (!Physics2D.OverlapBox(center, room.Tilemap.cellBounds.AsVector2() * 0.99f, 0))
+                    {
+                        valid = true;
+
+                        // Once room has spawned mark the old rooms connection as having been used.
+                        roomConnection.HasConnection = true;
+                    }
+                }
+
+                // Instantiate room in direction.
+                var roomInstance = Instantiate(room, center - room.Tilemap.cellBounds.center.AsVector2(), Quaternion.identity);
+
+                // Mark the new rooms connection with the old room.
+                foreach (RoomConnection connection in roomInstance.roomConnections)
```

Comment:
ideally we make roomConnections a dictionary<direction, bool>, so we can just look up the state in a direction instead of doing any loops